bindings/go/generator/ocmtypegen/main.go

```diff
@@ -1,114 +1,248 @@
-						}
-
-						for _, field := range structType.Fields.List {
-							if !strings.Contains(field.Tag.Value, `json:"type"`) {
-								continue
-							}
-							ident, ok := field.Type.(*ast.SelectorExpr)
-							if !ok {
-								return "", nil, fmt.Errorf("type expression found at %q but was not an import from runtime", typeSpec.Name.String())
-							}
-							if ident.Sel.Name == "type" {
-								return "", nil, fmt.Errorf("type expression found at %q but was unexported", typeSpec.Name.String())
-							}
-							if ident.Sel.Name == "Type" {
-								structNames = append(structNames, typeSpec.Name.Name)
-								break
-							}
-						}
-					}
+		if packageName == "" {
+			packageName = file.Name.Name
+		}
+
+		for _, decl := range file.Decls {
+			genDecl, ok := decl.(*ast.GenDecl)
+			if !ok || genDecl.Tok != token.TYPE {
+				continue
+			}
+
+			for _, spec := range genDecl.Specs {
+				typeSpec, ok := spec.(*ast.TypeSpec)
+				if !ok || !hasMarker(genDecl.Doc, typeSpec.Doc) {
+					continue
 				}
+
+				structType, ok := typeSpec.Type.(*ast.StructType)
+				if !ok || !hasRuntimeTypeField(structType) {
+					continue
+				}
+
+				typesToGenerate = append(typesToGenerate, typeSpec.Name.Name)
 			}
 		}
 	}
 
-	return node.Name.Name, structNames, nil
+	return packageName, typesToGenerate, nil
 }
 
-func main() {
-	// Get the source file path from command-line arguments
-	if len(os.Args) < 2 {
-		log.Fatal("Usage: //go:generate ocmtypegen $GOFILE")
+// findGoPackages recursively walks a directory to find all folders containing valid Go files.
+func findGoPackages(root string) ([]string, error) {
+	var packages []string
+
+	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+		if err != nil || !info.IsDir() {
+			return err
+		}
+		files, err := os.ReadDir(path)
+		if err != nil {
+			return err
+		}
+		for _, file := range files {
+			if !file.IsDir() && isValidGoFile(file.Name()) {
+				packages = append(packages, path)
+				break
+			}
+		}
+		return nil
+	})
+	return packages, err
+}
+
+// isValidGoFile checks if a file should be considered for parsing.
+func isValidGoFile(name string) bool {
+	return strings.HasSuffix(name, ".go") &&
+		!strings.HasSuffix(name, "_test.go") &&
+		!strings.HasPrefix(name, "zz_generated.")
+}
+
+// hasMarker returns true if any comment group contains the typegen marker.
+func hasMarker(groups ...*ast.CommentGroup) bool {
+	for _, g := range groups {
+		if g == nil {
+			continue
+		}
+		for _, c := range g.List {
+			if strings.Contains(strings.TrimSpace(c.Text), typegenMarker) {
+				return true
+			}
+		}
 	}
-	sourceFile := os.Args[1]
-	outputFile := filepath.Join(filepath.Dir(sourceFile), "zz_generated.type.ocm.software.go")
+	return false
+}
 
-	pkg, structs, err := ParseStructsWithGenerateDirective(sourceFile)
+// hasRuntimeTypeField checks if the struct has a field named Type of type runtime.Type.
+func hasRuntimeTypeField(s *ast.StructType) bool {
```

Comment:
This got to be the most one that could break. Especially if we rename. We should definitely leave a comment on the Type saying if you rename this make sure you rename the generator or something. WDYT?

Reply:
added a log statement for this case specifically